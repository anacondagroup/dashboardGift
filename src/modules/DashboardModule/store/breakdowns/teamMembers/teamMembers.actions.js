import {
  TEAM_MEMBERS_BREAKDOWN_LOAD_REQUEST,
  TEAM_MEMBERS_BREAKDOWN_LOAD_SUCCESS,
  TEAM_MEMBERS_BREAKDOWN_LOAD_FAILS,
  TEAM_MEMBERS_DOWNLOAD_REPORT_REQUEST,
  TEAM_MEMBERS_DOWNLOAD_REPORT_SUCCESS,
  TEAM_MEMBERS_DOWNLOAD_REPORT_FAIL,
} from './teamMembers.types';

export const teamMembersLoadRequest = options => ({
  type: TEAM_MEMBERS_BREAKDOWN_LOAD_REQUEST,
  payload: options,
});

export const teamMembersLoadSuccess = breakdown => ({
  type: TEAM_MEMBERS_BREAKDOWN_LOAD_SUCCESS,
  payload: breakdown,
});

export const teamMembersLoadFail = error => ({
  type: TEAM_MEMBERS_BREAKDOWN_LOAD_FAILS,
  payload: error,
});

export const teamMembersDownloadReportRequest = options => ({
  type: TEAM_MEMBERS_DOWNLOAD_REPORT_REQUEST,
  payload: options,
});
export const teamMembersDownloadReportSuccess = () => ({
  type: TEAM_MEMBERS_DOWNLOAD_REPORT_SUCCESS,
});
export const teamMembersDownloadReportFail = () => ({
  type: TEAM_MEMBERS_DOWNLOAD_REPORT_FAIL,
});
