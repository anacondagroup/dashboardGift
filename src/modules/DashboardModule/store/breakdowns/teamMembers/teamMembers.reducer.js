import * as R from 'ramda';
import { renameKeys } from '@alycecom/utils';

import { initialBreakdownState } from '../../../helpers/breakdownSelectors.helpers';
import { setLoading, setLoaded } from '../../../../../helpers/lens.helpers';

import {
  TEAM_MEMBERS_BREAKDOWN_LOAD_REQUEST,
  TEAM_MEMBERS_BREAKDOWN_LOAD_SUCCESS,
  TEAM_MEMBERS_BREAKDOWN_LOAD_FAILS,
  TEAM_MEMBERS_DOWNLOAD_REPORT_REQUEST,
  TEAM_MEMBERS_DOWNLOAD_REPORT_SUCCESS,
  TEAM_MEMBERS_DOWNLOAD_REPORT_FAIL,
} from './teamMembers.types';

export const initialState = {
  ...initialBreakdownState,
};

export const renameBreakdownKeys = renameKeys({
  id: 'id',
  first_name: 'firstName',
  last_name: 'lastName',
  avatar: 'avatar',
  gifts_accepted: 'giftsAccepted',
  gifts_sent: 'giftsSent',
  gifts_viewed: 'giftsViewed',
  meetings_booked: 'meetingsBooked',
});

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case TEAM_MEMBERS_BREAKDOWN_LOAD_REQUEST:
      return {
        ...state,
        ...setLoading(true, state),
      };
    case TEAM_MEMBERS_BREAKDOWN_LOAD_SUCCESS:
      return {
        ...state,
        ...R.compose(setLoading(false), setLoaded(true))(state),
        breakdown: action.payload.map(team => renameBreakdownKeys(team)),
      };
    case TEAM_MEMBERS_BREAKDOWN_LOAD_FAILS:
      return {
        ...state,
        ...setLoading(false, state),
        error: action.payload,
      };
    case TEAM_MEMBERS_DOWNLOAD_REPORT_REQUEST:
      return {
        ...state,
        isReportLoading: true,
      };
    case TEAM_MEMBERS_DOWNLOAD_REPORT_SUCCESS:
    case TEAM_MEMBERS_DOWNLOAD_REPORT_FAIL:
      return {
        ...state,
        isReportLoading: false,
      };
    default:
      return state;
  }
};
