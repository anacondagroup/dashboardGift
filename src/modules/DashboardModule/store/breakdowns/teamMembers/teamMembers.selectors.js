import * as R from 'ramda';
import { updateSearch } from '@alycecom/modules';
import { pickNotEmpty, renameKeys } from '@alycecom/utils';

import { getBreakdownsState } from '../breakdowns.selectors';
import { makeGetTeamById } from '../../../../../store/teams/teams.selectors';
import { makeGetCampaignById } from '../../../../../store/campaigns/campaigns.selectors';
import { createSelectors } from '../../../helpers/breakdownSelectors.helpers';

const teamMembersApiParams = R.pick([
  'teamId',
  'dateRangeFrom',
  'dateRangeTo',
  'sort',
  'sortDirection',
  'search',
  'campaignId',
]);

export const [
  getTeamMembersState,
  getTeamMembersBreakdown,
  getTeamMembersIsLoading,
  getTeamMembersReportIsLoading,
  getTeamMembersIsLoaded,
] = createSelectors(getBreakdownsState)('teamMembers');

export const getApiCallQuery = R.compose(
  R.curryN(2, updateSearch)(''),
  renameKeys({
    teamId: 'team_id',
    campaignId: 'campaign_id',
    sort: 'sort_column',
    sortDirection: 'sort_direction',
    search: 'search',
    dateRangeFrom: 'date_range_from',
    dateRangeTo: 'date_range_to',
  }),
  pickNotEmpty,
  teamMembersApiParams,
);

export const getReportFileName = apiParams => state => {
  let team = null;
  let date = '';
  let campaign = null;

  if (apiParams.teamId) {
    const { name: teamName } = makeGetTeamById(apiParams.teamId)(state);
    team = `-${teamName}-Team`;
  }

  if (apiParams.campaignId) {
    const { name: campaignName } = makeGetCampaignById(apiParams.campaignId)(state);
    campaign = `-${campaignName}-Moment`;
  }

  if (apiParams.dateRangeFrom) {
    date = `-${apiParams.dateRangeFrom}`;
  }

  if (apiParams.dateRangeTo) {
    date += `-${apiParams.dateRangeTo}`;
  }

  return R.replace(/\./g, '_', `Team members breakdown${team || ''}${campaign || ''}${date || ''}`).concat('.xlsx');
};

export const buildApiUrl = ({ teamId, ...rest }) =>
  `/enterprise/dashboard/breakdown/teams/${teamId}/members?${getApiCallQuery(rest)}`;

export const buildReportUrl = params => `/enterprise/report/team_breakdown?${getApiCallQuery(params)}`;
