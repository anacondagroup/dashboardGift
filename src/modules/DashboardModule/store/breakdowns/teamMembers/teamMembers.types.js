export const TEAM_MEMBERS_BREAKDOWN_LOAD_REQUEST = 'TEAM_MEMBERS_BREAKDOWN_LOAD_REQUEST';
export const TEAM_MEMBERS_BREAKDOWN_LOAD_SUCCESS = 'TEAM_MEMBERS_BREAKDOWN_LOAD_SUCCESS';
export const TEAM_MEMBERS_BREAKDOWN_LOAD_FAILS = 'TEAM_MEMBERS_BREAKDOWN_LOAD_FAILS';

export const TEAM_MEMBERS_DOWNLOAD_REPORT_REQUEST = 'TEAM_MEMBERS_DOWNLOAD_REPORT_REQUEST';
export const TEAM_MEMBERS_DOWNLOAD_REPORT_SUCCESS = 'TEAM_MEMBERS_DOWNLOAD_REPORT_SUCCESS';
export const TEAM_MEMBERS_DOWNLOAD_REPORT_FAIL = 'TEAM_MEMBERS_DOWNLOAD_REPORT_FAIL';
