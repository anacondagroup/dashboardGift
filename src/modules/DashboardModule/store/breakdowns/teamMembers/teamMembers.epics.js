import { ofType } from 'redux-observable';
import { catchError, map, mergeMap, withLatestFrom, tap, debounce, takeUntil } from 'rxjs/operators/index';
import { of, timer } from 'rxjs';

import { TEAM_MEMBERS_BREAKDOWN_LOAD_REQUEST, TEAM_MEMBERS_DOWNLOAD_REPORT_REQUEST } from './teamMembers.types';
import {
  teamMembersLoadSuccess,
  teamMembersLoadFail,
  teamMembersDownloadReportSuccess,
  teamMembersDownloadReportFail,
} from './teamMembers.actions';
import { getReportFileName, buildApiUrl, buildReportUrl } from './teamMembers.selectors';

export const loadTeamMembersBreakdownEpic = (action$, state$, { apiService }) =>
  action$.pipe(
    ofType(TEAM_MEMBERS_BREAKDOWN_LOAD_REQUEST),
    debounce(() => timer(300)),
    map(({ payload }) => buildApiUrl(payload)),
    mergeMap(url =>
      apiService.get(url).pipe(
        map(response => teamMembersLoadSuccess(response.members)),
        catchError(response => of(teamMembersLoadFail(response))),
        takeUntil(action$.ofType(TEAM_MEMBERS_BREAKDOWN_LOAD_REQUEST)),
      ),
    ),
  );

export const downloadReportEpic = (action$, state$, { apiService, downloadFile }) =>
  action$.pipe(
    ofType(TEAM_MEMBERS_DOWNLOAD_REPORT_REQUEST),
    map(action => [buildReportUrl(action.payload), action]),
    mergeMap(([url, action]) =>
      apiService.getFile(url).pipe(
        withLatestFrom(state$.pipe(map(getReportFileName(action.payload)))),
        tap(([blob, filename]) => {
          downloadFile(blob, filename);
        }),
        map(() => teamMembersDownloadReportSuccess()),
        catchError(() => [teamMembersDownloadReportFail()]),
      ),
    ),
  );

export const teamMembersEpic = [loadTeamMembersBreakdownEpic, downloadReportEpic];
