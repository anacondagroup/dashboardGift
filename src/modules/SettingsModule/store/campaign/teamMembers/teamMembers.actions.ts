import { createAction } from 'redux-act';

import { ITeamMember } from './teamMembers.types';

const PREFIX = 'SETTINGS/CAMPAIGN/GENERAL';

export const loadTeamMembersRequest = createAction<{ campaignId: number }>(`${PREFIX}/LOAD_TEAM_MEMBERS_REQUEST`);
export const loadTeamMembersSuccess = createAction<{ members: ITeamMember[] }>(`${PREFIX}/LOAD_TEAM_MEMBERS_SUCCESS`);
export const loadTeamMembersFail = createAction(`${PREFIX}/LOAD_TEAM_MEMBERS_FAIL`);

export const loadTeamAdminsRequest = createAction<{ campaignId: number }>(`${PREFIX}/LOAD_ADMINS_REQUEST`);
export const loadTeamAdminsSuccess = createAction<{ admins: ITeamMember[] }>(`${PREFIX}/LOAD_ADMINS_SUCCESS`);
export const loadTeamAdminsFail = createAction(`${PREFIX}/LOAD_ADMINS_FAIL`);

export const clearData = createAction(`${PREFIX}/CLEAR_DATA`);
