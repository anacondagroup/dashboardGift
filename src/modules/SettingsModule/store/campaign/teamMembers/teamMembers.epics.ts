import { Epic } from 'redux-observable';
import { catchError, map, switchMap } from 'rxjs/operators';
import { ofType } from '@alycecom/utils';

import {
  loadTeamAdminsFail,
  loadTeamAdminsRequest,
  loadTeamAdminsSuccess,
  loadTeamMembersFail,
  loadTeamMembersRequest,
  loadTeamMembersSuccess,
} from './teamMembers.actions';
import { ITeamMembersResponse } from './teamMembers.types';

const loadTeamMembersEpic: Epic = (action$, state$, { apiService, messagesService: { errorHandler } }) =>
  action$.pipe(
    ofType(loadTeamMembersRequest),
    switchMap(({ payload }) =>
      apiService.get(`/enterprise/dashboard/campaigns/${payload.campaignId}/team-members`).pipe(
        map(({ team_members: members }: ITeamMembersResponse) => loadTeamMembersSuccess({ members })),
        catchError(errorHandler({ callbacks: loadTeamMembersFail })),
      ),
    ),
  );

const loadTeamAdminsEpic: Epic = (action$, state$, { apiService, messagesService: { errorHandler } }) =>
  action$.pipe(
    ofType(loadTeamAdminsRequest),
    switchMap(({ payload }) =>
      apiService.get(`/enterprise/dashboard/campaigns/${payload.campaignId}/team-members?type=admin`).pipe(
        map(({ team_members: admins }: ITeamMembersResponse) => loadTeamAdminsSuccess({ admins })),
        catchError(errorHandler({ callbacks: loadTeamAdminsFail })),
      ),
    ),
  );

export default [loadTeamMembersEpic, loadTeamAdminsEpic];
