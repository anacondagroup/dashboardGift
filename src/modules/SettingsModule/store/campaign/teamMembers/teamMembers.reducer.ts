import { createReducer } from 'redux-act';

import {
  loadTeamMembersRequest,
  loadTeamMembersSuccess,
  loadTeamMembersFail,
  loadTeamAdminsRequest,
  loadTeamAdminsSuccess,
  loadTeamAdminsFail,
  clearData,
} from './teamMembers.actions';
import { ITeamMember } from './teamMembers.types';

export interface ITeamMembersState {
  members: ITeamMember[];
  admins?: ITeamMember[];
  isLoading: boolean;
}

export const initialTeamMembersState: ITeamMembersState = {
  members: [],
  admins: undefined,
  isLoading: false,
};

export const teamMembers = createReducer({}, initialTeamMembersState)
  .on(loadTeamMembersRequest, state => ({
    ...state,
    isLoading: true,
  }))
  .on(loadTeamMembersSuccess, (state, payload) => ({
    ...state,
    members: payload.members,
    isLoading: false,
  }))
  .on(loadTeamMembersFail, state => ({
    ...state,
    isLoading: false,
  }))

  .on(loadTeamAdminsRequest, state => ({
    ...state,
    isLoading: true,
  }))
  .on(loadTeamAdminsSuccess, (state, payload) => ({
    ...state,
    admins: payload.admins,
    isLoading: false,
  }))
  .on(loadTeamAdminsFail, state => ({
    ...state,
    isLoading: false,
  }))

  .on(clearData, () => ({
    ...initialTeamMembersState,
  }));
