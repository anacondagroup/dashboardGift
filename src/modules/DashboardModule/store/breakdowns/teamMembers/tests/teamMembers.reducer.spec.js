import { reducer, initialState } from '../teamMembers.reducer';
import { teamMembersLoadSuccess } from '../teamMembers.actions';

describe('team members reducer', () => {
  describe('teamsLoadSuccess', () => {
    it('Should transform breakdown data to app format', () => {
      const actual = reducer(
        initialState,
        teamMembersLoadSuccess([
          {
            avatar: 'https://s3.amazonaws.com/cdn-dev.alyce.co/images/users/u2_33201.jpg',
            id: 2,
            first_name: 'Greg',
            last_name: 'Segall',
            gifts_sent: 0,
            gifts_viewed: 0,
            gifts_accepted: 0,
            meetings_booked: 0,
          },
          {
            avatar: 'http://alyce.site/images/avatar.png',
            id: 13,
            first_name: 'Valentin',
            last_name: 'Belorustcev',
            gifts_sent: 0,
            gifts_viewed: 0,
            gifts_accepted: 0,
            meetings_booked: 0,
          },
        ]),
      );

      expect(actual).toMatchSnapshot();
    });
  });
});
