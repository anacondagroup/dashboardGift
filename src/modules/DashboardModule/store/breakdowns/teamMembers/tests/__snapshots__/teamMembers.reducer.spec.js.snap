// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`team members reducer teamsLoadSuccess Should transform breakdown data to app format 1`] = `
Object {
  "breakdown": Array [
    Object {
      "avatar": "https://s3.amazonaws.com/cdn-dev.alyce.co/images/users/u2_33201.jpg",
      "firstName": "Greg",
      "giftsAccepted": 0,
      "giftsSent": 0,
      "giftsViewed": 0,
      "id": 2,
      "lastName": "Segall",
      "meetingsBooked": 0,
    },
    Object {
      "avatar": "http://alyce.site/images/avatar.png",
      "firstName": "Valentin",
      "giftsAccepted": 0,
      "giftsSent": 0,
      "giftsViewed": 0,
      "id": 13,
      "lastName": "Belorustcev",
      "meetingsBooked": 0,
    },
  ],
  "error": null,
  "isLoaded": true,
  "isLoading": false,
  "isReportLoading": false,
  "pagination": Object {
    "current_page": 1,
  },
}
`;
